### PR TITLE
SERVER-17832 Fix Resource Leak on SSL Connect and Accept Exception

### PR DESCRIPTION
**Pull request by guoyr**

### CODE REVIEW

File: src/mongo/util/net/ssl_manager.cpp [hunk 1]
```diff
@@ -831,39 +832,31 @@ namespace mongo {
     }
 
     SSLConnection* SSLManager::connect(Socket* socket) {
-        SSLConnection* sslConn = new SSLConnection(_clientContext, socket, NULL, 0);
-        ScopeGuard sslGuard = MakeGuard(::SSL_free, sslConn->ssl);
-        ScopeGuard bioGuard = MakeGuard(::BIO_free, sslConn->networkBIO);
+        std::unique_ptr<SSLConnection> sslConn = stdx::make_unique<SSLConnection>(_clientContext, socket, (const char*)NULL, 0);
```
Comment (acmorrow) [inline src/mongo/util/net/ssl_manager.cpp]: The JIRA issue that you referenced indicates that this fix is intended for backport to pre-3.1 branches, where C++11 is not permitted.

Comment (guoyr) [inline src/mongo/util/net/ssl_manager.cpp]: Understood. The consensus from the CR is that there needs to be another version for the backport. I've submitted another PR on the 3.0 branch which uses auto_ptr instead. https://github.com/mongodb/mongo/pull/950
